Continuous Profiling - ddprof updates for 0.9.0 (#14783)
DataDog/documentation: content/en/tracing/profiler/enabling/ddprof.md

---
title: Enabling the Profiler for Compiled Languages
kind: Documentation
code_lang: ddprof
type: multi-code-lang
code_lang_weight: 90
aliases:

>   - /tracing/profiler/enabling/linux/

Reviewer: Just to make sure:  this alias exists to make sure old links stay active?

Reviewer 2: Correct

---

<div class="alert alert-warning">

> The compiled languages profiler, <code>ddprof</code> is in public beta. Datadog recommends evaluating the profiler in a non-sensitive environment before deploying in production.

Reviewer 2: ```suggestion
The profiler for compiled languages, <code>ddprof</code> is in public beta. Datadog recommends evaluating the profiler in a non-sensitive environment before deploying in production.
```

</div>

> The Datadog Profiler for compiled languages (`ddprof`) uses OS level APIs to collect profiling data. It is ideally suited for applications written in compiled languages, such as C, C++, or Rust.

Reviewer: What's the product name here?
* Datadog Profiler
* Datadog Profiler for Compiled Languages

I'd guess the latter and suggest that the casing should follow, but I actually have no idea!  I think this is the first Profiling product where the target (runtime?) wasn't already a proper name!

Reviewer 2: Ah, the name. I believe the official name of the entire feature/product is (Datadog) Continuous Profiler. And the specific language libraries each contain a profiler _for lang_ feature that you can enable. which we can also refer to as the _lang_ profiler (or sometimes _lang_ Profiler). This particular profiler is not alone in being all over the map in terms of capitalization and naming used in the docs. So, throughout these docs I have tried to keep the nouns common (uncapitalized) except in H1/navigation where we Title Case things. So, I would call these the docs for the "profiler for compiled languages" (no caps). I'll do a spot check for consistency here, but getting consistency across the whole docs is a bigger project :)

Reviewer: Gotcha!  Thanks for the clarification!

> The Datadog Profiler for compiled languages (`ddprof`) uses OS level APIs to collect profiling data. It is ideally suited for applications written in compiled languages, such as C, C++, or Rust.

Reviewer 2: ```suggestion
The Datadog Continuous Profiler for compiled languages (`ddprof`) uses OS level APIs to collect profiling data. It is ideally suited for applications written in compiled languages, such as C, C++, or Rust.
```

## Requirements

Supported operating systems

> : Linux kernel v4.17+, libc 2.23+

Reviewer: Looking at the current version of ddprof, we actually support glibc 2.18+.

For the curious:
```
$ readelf -sW ./ddprof | grep GLIBC_ | sed 's/.*GLIBC/GLIBC/g' | sort | uniq
GLIBC_2.10
GLIBC_2.10 (11)
GLIBC_2.14
GLIBC_2.14 (4)
GLIBC_2.15
GLIBC_2.15 (18)
GLIBC_2.16
GLIBC_2.16 (14)
GLIBC_2.18
GLIBC_2.18 (17)
<snip>
```

Author: We could indeed add a test in the CI / releasing to make sure we never exceed 2,18. For now the guarantee was around 2.23 (due to build infra). Considering you validated this for the current release, it is worth mentioning it.

Author: Too bad we don't naturally have centos 7 covered ! (2.17 afaik)

Supported architecture
: `amd64` or `arm64` processors

Serverless
: `ddprof` is not supported on serverless platforms, such as AWS Lambda.

OS Settings
: `perf_event_paranoid` kernel setting is 2 or less (see [Troubleshooting][1])

Debugging information

> : Symbols should be available

Reviewer: I think we should be more specific by saying something like, "ddprof will not be able to provide human-readable function names if the symbol table is stripped."  This communicates two things

* The user may not be aware that "symbols should be available" means "the ELF symbol table must be present in the main binary."  However, they may be aware that they "strip" their executables to save room (this is becoming more and more common).  Moreover, I think we should speak about the thing (the symbol table) rather than the information it provides (symbol names).
* ddprof kinda works if the symbol table is missing, it's just that the resulting flamegraphs are annotated with execution addresses rather than symbol names.  Pointing out that the user won't see human-readable symbol names also allows them to self-serve in some of the diagnostics.

WDYT?

Reviewer 2: I added that as an additional clarification on the existing sentence. Let me know what you think.

Author: Thanks, this is indeed more and more common. It was worth adding a sentence.

> : Symbols should be available

Reviewer 2: ```suggestion
: Symbols should be available. The profiler cannot provide human-readable function names if the symbol table is stripped.
```

```

{{% /tab %}}

> {{% tab "Parameters" %}}

Reviewer: Ah yeah, good change.

{{< /tabs >}}

> 5. A minute or two after starting your application, your profiles appear on the [Datadog APM > Profiler page][3].

Reviewer: Is "a minute or two" still accurate in the contemporary product?  I know that some APM customers complain because it can take ~10 minutes to get the first trace.  Want to be sure we set expectations properly in the average case.

(FWIW, I usually only have to wait a minute or two, just want to make sure that's what our customers can reliably expect)

Author: Looking at lag time, we are indeed up to 4 minutes (when everything is OK). The usual wait time will be around 2/3 minutes.

Author: It is still very relevant to tell the user to wait for a few minutes.

### Globalmode

Global mode is intended for debug purposes. When `global` is set to `yes`, the profiler attempts to profile all the visible processes. 
This requires elevated permissions (for example, running as root or granting `CAP_PERFMON`, `CAP_SYSADMIN`) or setting `perf_event_paranoid` to `-1`.

    ```bash
    ./ddprof --environment staging --global yes --service_version full-host-profile
    ```

> When the profiler has root UID, all visible processes are instrumented.

Reviewer: "has root UID" -> "is run as the root user" ?

> When the profiler has root UID, all visible processes are instrumented.

Reviewer 2: ```suggestion
When the profiler is run as the root user, all visible processes are instrumented.
```
